### PR TITLE
Ghost icons fixes 2

### PR DESCRIPTION
**Pull request by saurtron**
### Work done

- Fix dead ghost icons appearing for specs.
  - they don't see the ghosts so shouldn't see the icons either
- Fix "dead" and "alive but not under radar" ghost icons being indistinguishable from  regular radar icons.
  - achieved this by making them show more dimmed
- Make the dimming configurable through UnitGhostIconsDimming config option
  - Also setting to 0 allows disabling icons for ghosts not under radar (first step to fix ([#641](https://github.com/beyond-all-reason/spring/issues/641))
- Apply some cosmetic changes for consistency between DrawUnitMiniMapIcons and DrawUnitIconsScreen code.

### Related issues:

- https://github.com/beyond-all-reason/spring/issues/2031
- https://github.com/beyond-all-reason/spring/issues/2032
- https://github.com/beyond-all-reason/spring/issues/641

### Screenshots

how specs see it (click to see larger):

how a player sees it:

(note how the two solars at the bottom are dimmed, one is dead and the other one alive, but the player doesn't know since he doesn't have radar coverage).

all colors comparison for 16 players (dimming 0.5):

team-red

team-blue

on minimap:

under-radar

### CODE REVIEW

File: rts/Rendering/Units/UnitDrawer.cpp [hunk 1]
```diff
@@ -631,19 +644,35 @@ void CUnitDrawerGLSL::DrawUnitIconsScreen() const
 
 			// use white for selected units
 			SColor color = unit->isSelected ? color4::white : SColor{ teamHandler.Team(unit->team)->color };
+			if (!gu->spectatingFullView && !(unit->losStatus[gu->myAllyTeam] & LOS_INRADAR)) {
```
Comment (sprunk) [inline rts/Rendering/Units/UnitDrawer.cpp:647]: In theory a unit under a ghost can be selected too.

File: rts/Rendering/Units/UnitDrawer.cpp [hunk 2]
```diff
@@ -385,6 +387,7 @@ void CUnitDrawerGLSL::DrawUnitMiniMapIcons() const
 				continue;
 
 			const uint8_t* color = &defaultColor[0];
+			SColor useColor;
```
Comment (sprunk) [inline rts/Rendering/Units/UnitDrawer.cpp:390]: Does this need to be a separate var? Seems it's just used as a temporary in line 411 but could just do `color = SColor{color[0]* dim, color[1] * dim ...}` same as line 431, or `color.rgb *= dim` same as lines 650-652.

File: rts/Rendering/Units/UnitDrawer.cpp [hunk 3]
```diff
@@ -631,19 +644,35 @@ void CUnitDrawerGLSL::DrawUnitIconsScreen() const
...
-			SColor color = SColor{ teamHandler.Team(ghost->team)->color };
+				pos = camera->CalcViewPortCoordinates(pos);
+				if (pos.z > 1.0f || pos.z < 0.0f)
+					continue;
 
-			DrawUnitIconScreen(rb, icon, pos, color, ghost->radius, false);
+				SColor color = SColor{ teamHandler.Team(ghost->team)->color };
+				color.r = color.r*ghostIconDimming;
+				color.g = color.g*ghostIconDimming;
+				color.b = color.b*ghostIconDimming;
```
Comment (sprunk) [inline rts/Rendering/Units/UnitDrawer.cpp:672]: consider
```suggestion
				color.r *= ghostIconDimming;
				color.g *= ghostIconDimming;
				color.b *= ghostIconDimming;
```

File: rts/Rendering/Units/UnitDrawer.cpp [hunk 4]
```diff
@@ -402,6 +405,12 @@ void CUnitDrawerGLSL::DrawUnitMiniMapIcons() const
 					color = teamHandler.Team(unit->team)->color;
 				}
 			}
+			if (!gu->spectatingFullView && !(unit->losStatus[gu->myAllyTeam] & LOS_INRADAR)) {
```
Comment (sprunk) [inline rts/Rendering/Units/UnitDrawer.cpp:408]: Optionally would be good if there was some sort of `gu->IsUnitVisibleAsGhost(unit)` or `unit->IsVisibleAsGhost()` or such helper. Not necessarily in the scope of this patch though

**Review by sprunk — COMMENTED** [review]
> Fix "dead" and "alive but not under radar" ghost icons being indistinguishable from regular radar icons.
> achieved this by making them show more dimmed

This doesn't fix it because you can have dark teamcolors where players will struggle to tell the difference (incl. completely black), and by producing a different colour the ghosts can become indistinguishable from another player's units instead (e.g. in a 2v2v2v2v... team FFA the teams are often red+darkred vs green+darkgreen vs blue+darkblue etc).

I'm happy that 0.0 exists though and even dimming is probably a positive tradeoff so it's fine.

File: rts/Rendering/Units/UnitDrawer.cpp (hunk 3 above)
Comment (saurtron) [reply rts/Rendering/Units/UnitDrawer.cpp]: indeed, applied it

File: rts/Rendering/Units/UnitDrawer.cpp (hunk 2 above)
Comment (saurtron) [reply rts/Rendering/Units/UnitDrawer.cpp]: changed SColor use to a different way so I think that's now applied but in a different way (uses a separate var but out of the loop).

didn't like the color initializer, but pls check again and see what you think.

File: rts/Rendering/Units/UnitDrawer.cpp [hunk 5]
```diff
@@ -402,6 +405,12 @@ void CUnitDrawerGLSL::DrawUnitMiniMapIcons() const
...
+				if (!ghostIconDimming)
```
[the remark this answers is not in the document]
Comment (saurtron) [reply rts/Rendering/Units/UnitDrawer.cpp]: Changed this one to `use == 0.0` instead, let me know if you still want me doing one const at the beginning. I guess it can save a few ops, so likely good to do, but just to be sure.

File: rts/Rendering/Units/UnitDrawer.cpp (hunk 1 above)
Comment (saurtron) [reply rts/Rendering/Units/UnitDrawer.cpp:649]: Hmm, in this case I'm not sure it makes a difference... if it's in radar the test won't apply, and if not in radar I think you would also either have fullView or also wouldn't be possible to select it (at least I tried god mode, and it does let me select enemy units, but when losing radar I lose it from the selection too (unless I have fullView too).

File: rts/Rendering/Units/UnitDrawer.cpp (hunk 4 above)
Comment (saurtron) [reply rts/Rendering/Units/UnitDrawer.cpp:407]: Maybe for a future patch then, note also it's not just about it being visible but also about the radar status, ie the method can be added, but not necessarily directly related to this test.

File: rts/Rendering/Units/UnitDrawer.cpp (hunk 1 above)
Comment (sprunk) [reply rts/Rendering/Units/UnitDrawer.cpp:649]: Try `/specfullview 2`

**Comment by saurtron** [comment]
> > Fix "dead" and "alive but not under radar" ghost icons being indistinguishable from regular radar icons.
> > achieved this by making them show more dimmed
> 
> This doesn't fix it because you can have dark teamcolors where players will struggle to tell the difference (incl. completely black), and by producing a different colour the ghosts can become indistinguishable from another player's units instead (e.g. in a 2v2v2v2v... team FFA the teams are often red+darkred vs green+darkgreen vs blue+darkblue etc).
> 
> I'm happy that 0.0 exists though and even dimming is probably a positive tradeoff so it's fine.

Added some images to the description, with comparison of all basic 16 colors as used in bar, indeed with so many colors some cases could still be difficult to tell apart, imo the biggest offender is red/brown, but it'll depend on person and monitor. I think it will generally help, but yeah it's not perfect.

Maybe in the future an alternative approach could be tried, like maybe making them outline only for icons not under radar or so, but I think that'll take some more work, maybe someone has an even better idea.

File: rts/Rendering/Units/UnitDrawer.cpp (hunk 1 above)
Comment (saurtron) [reply rts/Rendering/Units/UnitDrawer.cpp:649]: > Try `/specfullview 2`

hmm, trying but can't make that one do anything, do I need to activate something else before that?

Comment (sprunk) [reply rts/Rendering/Units/UnitDrawer.cpp:649]: Sounds fine to keep as-is if it's not actually trivial to reproduce, support for this edge case can be trickled in later.

**Review by sprunk — APPROVED** [review]
Looks good, will test later.

**Comment by sprunk** [comment]
> Maybe in the future an alternative approach could be tried, like maybe making them outline only for icons not under radar or so, but I think that'll take some more work, maybe someone has an even better idea.

The proper approach is to let Lua draw the icons / specify a shader. Then Lua can apply any logic it wants, whether that's drawing just the outline, or some smart tailored dimming that avoids other existing colors and adjusts intensity depending on how dark the original color is, or flashing/shrinking/inverting/whatever.

This also applies to ghosts as a whole.

File: rts/Rendering/Units/UnitDrawer.cpp (hunk 1 above)
Comment (saurtron) [reply rts/Rendering/Units/UnitDrawer.cpp:650]: > Sounds fine to keep as-is if it's not actually trivial to reproduce, support for this edge case can be trickled in later.

Well, it's not so much about reproducing, as to what is really supported... after a bit more testing I think engine does support this, although game(s) probably not.

Then, it's about if we think a selected unit's icon should be dimmed when not under los/radar... a bit of a hard call, on one hand it's a pretty niche use at least as games using this atm (not even sure engine supports this use case 100%, tbh), on the other if supported pbbly not dimming is better.

Anyways since it looks like it's supported I made it not dim the selection for now (ie, **fixed** it), since in that case I find grey to be a bit confusing...

@lhog if you can read this and chime in, would be good as I'm not 100% on this one, but if engine should support it I think we should support it as good as possible.

For reference this is how it shows when dimmed:

dimmedselection2

Comment (lhog) [reply rts/Rendering/Units/UnitDrawer.cpp:650]: Agree, the selected units shouldn't dim.

**Comment by saurtron** [comment]
Did a number of smaller changes, mostly for code consistency between DrawUnitMiniMapIcons and DrawUnitIconsScreen (did fix a small discrepancy on "selected" icon rendering between them too).

File: rts/Rendering/Units/UnitDrawer.cpp [hunk 6]
```diff
@@ -357,6 +357,7 @@ void CUnitDrawerGLSL::DrawUnitMiniMapIcons() const
 	sh.SetUniform("alphaCtrl", 0.0f, 1.0f, 0.0f, 0.0f); // GL_GREATER > 0.0
 
 	SColor currentColor;
+	const auto allyTeam = gu->myAllyTeam;
```
Comment (sprunk) [inline rts/Rendering/Units/UnitDrawer.cpp:360]: I'd call this `myAllyTeam` because this is UnitDrawer so it could plausibly be the unit's allyteam.

Comment (saurtron) [reply rts/Rendering/Units/UnitDrawer.cpp:360]: yeah, me too, just kept the previous name since it was already defined elsewhere, but will change it in both places then.

Comment (saurtron) [reply rts/Rendering/Units/UnitDrawer.cpp]: ok, done
